### PR TITLE
Select/reset button for events

### PR DESCRIPTION
**Pull request by OliverBalfour**

### CODE REVIEW

**Comment by OliverBalfour** [comment]
Only one regression I found: the agenda section has missing data. I've just fixed this

**Comment by pl4nty** [comment]
I'm out this morning, but I'll get to this today, thanks Oliver! Would normally avoid global state/rerenders for performance and complexity, but seems fine here (even faster than the FC API anecdotally) and I don't see the state growing too much.

File: src/Export.js [hunk 1]
```diff
@@ -17,7 +17,7 @@ const Export = ({ API }) => {
       <SiApple /> WebCal (eg iOS)
     </Dropdown.Item>
     <Dropdown.Item eventKey="gcal" target={"_blank"} rel={"noreferrer"} 
-      href={`https://www.google.com/calendar/render?cid=${webcal}`}>
+      href={`https://www.google.com/calendar/render?cid=${encodeURIComponent(webcal.replace('https', 'http'))}`}>
```
Comment (pl4nty) [inline src/Export.js:20]: Dug a bit deeper and I'll need to refactor the link generation (to fix Exchange/Outlook), but I'll do that in another branch

File: src/Toolbar.js [hunk 2]
```diff
@@ -1,142 +1,15 @@
...
-  useEffect(() => {
-    m.forEach(([module, occurrences]) => occurrences.split(',').forEach(o => {
-      if (!o || !selectedModules.map(({ id }) => id).includes(module)) return
-      const r = o.match(/([^0-9]*)([0-9]+)$/)
-      selectOccurrence(calendar, module, r[1], parseInt(r[2]))
-    }))
-  }, [m, selectedModules, calendar])
 
+export default forwardRef(({ API, state: {
+  timeZone, year, session, sessions, JSON, modules, selectedModules,
```
Comment (pl4nty) [inline src/Toolbar.js:10]:
```suggestion
  timeZone, year, session, sessions, timetableData, modules, selectedModules,
```

File: src/Toolbar.js [hunk 3]
```diff
@@ -1,142 +1,15 @@
...
+  setTimeZone, setYear, setSession, setSessions, setJSON, setModules, setSelectedModules,
```
Comment (pl4nty) [inline src/Toolbar.js:11]:
```suggestion
  setTimeZone, setYear, setSession, setSessions, setTimetableData, setModules, setSelectedModules,
```

File: src/Calendar.js [hunk 4]
```diff
@@ -182,7 +103,7 @@ export default forwardRef((props, ref) => {
         listDayFormat: { weekday: 'long', month: 'short', day: 'numeric' },
         displayEventTime: true,
         weekends: true,
-        eventContent: formatEventContent
+        eventContent: e => formatEventContent(state)(e)
```
Comment (pl4nty) [inline src/Calendar.js:106]:
```suggestion
        ...customEvents
```

File: src/App.js [hunk 5]
```diff
@@ -17,13 +18,121 @@ let App = () => {
...
+
   // fluid="xxl" is only supported in Bootstrap 5
   return <Container fluid>
     <h2 className="mt-2">ANU Timetable</h2>
 
-    <Toolbar API={API} ref={calendar} timeZone={timeZone} setTimeZone={setTimeZone} />
+    <Toolbar API={API} ref={calendar} state={state} />
 
-    <Calendar API={API} ref={calendar} timeZone={timeZone} />
+    <Calendar API={API} ref={calendar} state={state} />
```
Comment (pl4nty) [inline src/App.js:135]:
```suggestion
    <Calendar ref={calendar} state={state} />
```

**Review by pl4nty — CHANGES_REQUESTED** [review]
Not sure how I feel about the state handling, but it works and the rest is great so I'm happy.

File: src/index.css [hunk 6]
```diff
@@ -155,14 +155,14 @@
 
 /* apply default colour to things which shouldn't be links */
 .fc .fc-event-main a {
-  color: var(--fc-event-text-color, #fff) !important;
+  color: var(--fc-event-text-color, inherit) !important;
 }
 
 /* apply override to actual links in events */
 .fc .fc-event a {
```
[the remark this answers is not in the document]
Comment (OliverBalfour) [reply src/index.css:162]: If they're using links internally, perhaps? As we're forcing the color to be inherited we can always change the bootstrap button's parent's color but that's a little nasty I suppose

File: src/Calendar.js [hunk 7]
```diff
@@ -25,115 +25,36 @@ rrulePlugin.recurringTypes[0].expand = function (errd, fr, de) {
...
   // fix PR is unmerged since Apr 2021: fullcalendar/fullcalendar#5710
-  const { location, locationID, lat, lon } = event.extendedProps
-  const url = lat ? `https://www.google.com/maps/search/?api=1&query=${lat},${lon}` : locationID
   const locationLine = url
     ? <a href={url} target="_blank" rel="noreferrer">{location}</a>
     : location;
+  const dispatch = f => f(event.source.id, event.groupId, event.extendedProps.occurrence)
+  const button = activity.startsWith('Lec') ? null :
+    hasMultipleOccurrences
+      ? <button className='choose-button' onClick={() => dispatch(selectOccurrence)}>Choose</button>
```
[the remark this answers is not in the document]
Comment (OliverBalfour) [reply src/Calendar.js:39]: I prefer the text link buttons because sometimes the background colour clashes with the button colour - thoughts?

**Comment by pl4nty** [comment]
Interesting bug with the text overflow, also affects buttons

image

**Comment by OliverBalfour** [comment]
Re: state handling, the reason I moved everything into App was because a lot of stuff didn't make sense inside Toolbar, eg which events are selected, and I needed a way to propagate that info into the Calendar. The state handling is not ideal, I agree, I only did precisely enough refactoring to enable me to pass the state where it was needed, but didn't clean it up afterwards.

Virtually all of the state handling code is the same or very similar, it's just been moved and is all passed down

**Comment by OliverBalfour** [comment]
> Interesting bug with the text wrapping, also affects buttons
> 
> image

Good catch, just fixed (missing `overflow-y: hidden`)

**Comment by pl4nty** [comment]
All good, originally I stored all user input in the toolbar (and using the calendar ref just to display), but that fell apart when the calendar itself took input. Glad we're on the same page, not ideal as-is but a big improvement on what I had, and enough to release to users.

**Comment by pl4nty** [comment]
Nice! TIL that exists

**Comment by OliverBalfour** [comment]
Thinking about the state management, the main reason it's ugly is because CalendarApi is imperative and doesn't let you pass an events list prop. Longer term, I might make a FullCalendar wrapper that simply adds an `events` prop that wrangles CalendarApi for you. Then the only complex bits is generating the list of events to pass via props - that's most of the contents of the giant `useEffect` - but it could be moved to a helper function in `Calendar.js` when we can pass events as a prop

**Comment by pl4nty** [comment]
ahhhhh, I thought FC was imperative? never tested cause I was attempting to wrangle the API, which was a poor choice in hindsight. https://fullcalendar.io/docs/initialEvents

**Comment by OliverBalfour** [comment]
> ahhhhh, I thought FC was imperative? never tested cause I was attempting to wrangle the API, which was a poor choice in hindsight. https://fullcalendar.io/docs/initialEvents

Yeah what I'm proposing is a wrapper that makes a prop like `initialEvents`, except it actually *does* update the events when the prop changes. It'd be a separate file that does some funky stuff with effects to translate changes to the prop into `CalendarApi` calls, so it abstracts away the imperative-ness and presents a declarative API

Edit: see your edit now, gotcha

File: src/Calendar.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (pl4nty) [reply src/Calendar.js:39]: Agreed, I'm testing locally to see if the -outline variants or links fix that. Happy to merge and PR button stuff later.

File: src/index.css (hunk 6 above)
[the remark this answers is not in the document]
Comment (pl4nty) [reply src/index.css:162]: Will test this, variant=link uses links internally but I'm sure there's a way to exclude them

**Comment by pl4nty** [comment]
Yeah - I think the default events prop (eg https://fullcalendar.io/docs/events-array) is declarative, and initialEvents is imperative for those that want it

**Review by pl4nty — APPROVED** [review]
LGTM ❤

**Pull request merged**